### PR TITLE
Fix memory leak in Base64Encode

### PR DESCRIPTION
**Pull request by Jinming-Hu**
closes https://github.com/Azure/azure-sdk-for-cpp/issues/2426

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

**Review by vhvb1989 — APPROVED** [review]
Does it mean that BIO_NOCLOSE was preventing bufferPtr to be free?  or did we have 2 issues here, one for using bufferPtr after free and another for using BIO_NOCLOSE?

**Review by ahsonkhan — APPROVED** [review]

**Comment by ahsonkhan** [comment]
> Does it mean that BIO_NOCLOSE was preventing bufferPtr to be free? 

Good question. If that's the case, then did we not have an actual use after free issue from https://github.com/Azure/azure-sdk-for-cpp/pull/2409? Or is the fact that there is an openssl bug that @antkmsft  pointed earlier where the buffer is still getting reed imply, the use-after-free bug existed - https://github.com/openssl/openssl/issues/10448?

I am not very familiar with the open ssl bio APIs yet, @Jinming-Hu or @jorgen do you know?

**Comment by jorgen** [comment]
Look at the function. The way I see it, the way the function was written there could be two side effects. One was that all memory allocated by the BIO calls whered freed by the BIO_free_all function, resulting in the subsequent copy being "use after free". The other is that not all memory from BIO is freed and either there has to be a cleanup thread somewhere (this would be very surprising) or this function is leaking memory. After reading the documentation briefly I desided that there was a high probabillity of "use after free" and moved the copy a few lines up. If this wasn't the case it wouldn't matter because the function would still work as expected, but from looking at it it surly look like it wasn't right. Remember, I'm not using this function, I just encountered it when fixing compilation issues.

Btw, I think the close flag is a red herring, and its intended use is when the BIO structure is used with a stream to stdout or some other stream opened outside the BIO calls.

File: sdk/core/azure-core/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ## 1.1.0-beta.1 (Unreleased)
 
+### Bug Fixes
+
+- Fixed a memory leak issue in `Base64Encode()`.
```
[the remark this answers is not in the document]
Comment (Jinming-Hu) [reply sdk/core/azure-core/CHANGELOG.md:8]: https://github.com/Azure/azure-sdk-for-cpp/issues/2444

**Pull request merged**